### PR TITLE
test(wallet-cli): improve device part test coverage [LIVE-28271]

### PR DESCRIPTION
**Pull request by Justkant**
### ✅ Checklist

- [ ] `npx changeset` was attached. only tests
- [x] **Covered by automatic tests.** 
- [x] **Impact of the changes:**
  - device part of the wallet-cli

### 📝 Description

This pull request introduces significant improvements to device connection management and testing infrastructure for the wallet CLI, with a focus on robustness, error handling, and concurrency safety. The changes ensure that device sessions are correctly reset or disposed of on failure, prevent race conditions during concurrent connection attempts, and enhance test coverage for edge cases in device management. Below are the most important changes:

**Device Connection Robustness and Error Handling**

* Refactored the device connection flow in `connect-ledger-app.ts` to use a discriminated union return type (`ConnectAppOnceResult`) for clearer success/error handling, and improved timeout/cancellation logic to ensure resources are cleaned up on silence timeouts. [[1]](diffhunk://#diff-dcd903b24b4f39765a6cdead83a71c02d30e806ae2969451ead6b8af029e9f1aR22-R25) [[2]](diffhunk://#diff-dcd903b24b4f39765a6cdead83a71c02d30e806ae2969451ead6b8af029e9f1aL96-R102) [[3]](diffhunk://#diff-dcd903b24b4f39765a6cdead83a71c02d30e806ae2969451ead6b8af029e9f1aL108-R118) [[4]](diffhunk://#diff-dcd903b24b4f39765a6cdead83a71c02d30e806ae2969451ead6b8af029e9f1aL129-R133) [[5]](diffhunk://#diff-dcd903b24b4f39765a6cdead83a71c02d30e806ae2969451ead6b8af029e9f1aL142-R157) [[6]](diffhunk://#diff-dcd903b24b4f39765a6cdead83a71c02d30e806ae2969451ead6b8af029e9f1aL163-R195) [[7]](diffhunk://#diff-dcd903b24b4f39765a6cdead83a71c02d30e806ae2969451ead6b8af029e9f1aL194-R212)
* Improved error messages and error types in `register-dmk-transport.ts`, including introducing a custom `InitialSessionBusyError` and constants for user-facing messages, for better diagnostics and user guidance. [[1]](diffhunk://#diff-d95e1bf9e08e238e8fd5b165a7aaf9c62979b9641d328212f7f32fa2550d96b2R29-R32) [[2]](diffhunk://#diff-d95e1bf9e08e238e8fd5b165a7aaf9c62979b9641d328212f7f32fa2550d96b2R43-R58) [[3]](diffhunk://#diff-d95e1bf9e08e238e8fd5b165a7aaf9c62979b9641d328212f7f32fa2550d96b2L154-R180) [[4]](diffhunk://#diff-d95e1bf9e08e238e8fd5b165a7aaf9c62979b9641d328212f7f32fa2550d96b2L175-R223)

**Concurrency and Resource Management**

* Ensured that only one in-flight connection attempt is allowed at a time by introducing a `pendingTransport` promise, preventing race conditions when multiple concurrent calls to establish a connection are made. [[1]](diffhunk://#diff-d95e1bf9e08e238e8fd5b165a7aaf9c62979b9641d328212f7f32fa2550d96b2R43-R58) [[2]](diffhunk://#diff-d95e1bf9e08e238e8fd5b165a7aaf9c62979b9641d328212f7f32fa2550d96b2L175-R223)
* Updated teardown logic to fully await asynchronous teardown operations (like `destroyTransport` and `disconnect`) during reset and dispose flows, ensuring that resources are not leaked and that subsequent operations can proceed safely.

**Transport Layer Improvements**

* Added logic in `NodeWebUsbTransport` to track and share in-progress device connection setups, so that concurrent requests for the same device do not trigger redundant connection attempts, improving efficiency and correctness. [[1]](diffhunk://#diff-50bfcccc4bd51137b668895b26e650ef59eb2b27fc063e4ad419e968bc6a0e68R143-R146) [[2]](diffhunk://#diff-50bfcccc4bd51137b668895b26e650ef59eb2b27fc063e4ad419e968bc6a0e68R586-R595) [[3]](diffhunk://#diff-50bfcccc4bd51137b668895b26e650ef59eb2b27fc063e4ad419e968bc6a0e68L622-L630)

**Testing Enhancements**

* Extended the test suite in `register-dmk-transport.test.ts` to cover new edge cases, including persistent state resets after failures, concurrent connection attempts, and teardown behavior, using new helper utilities like `deferred` and `waitOneTick` for precise async control. [[1]](diffhunk://#diff-2302b800f7dd1ba8086697331d0e34d164da83076b2dd541a1727d81bce1ff82R19-R24) [[2]](diffhunk://#diff-2302b800f7dd1ba8086697331d0e34d164da83076b2dd541a1727d81bce1ff82R39-R86) [[3]](diffhunk://#diff-2302b800f7dd1ba8086697331d0e34d164da83076b2dd541a1727d81bce1ff82R132-R321)

These changes collectively make device management in the CLI more robust, reliable, and maintainable, especially under error conditions and concurrent usage scenarios.

### ❓ Context

- **JIRA or GitHub link**: [LIVE-28271]
- **ADR link (if any)**: 

---

### 🧐 Checklist for the PR Reviewers

- **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- **Any new dependencies** have been justified and documented.
- **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

[LIVE-28271]: https://ledgerhq.atlassian.net/browse/LIVE-28271?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by gre-ledger — APPROVED** [review]

File: apps/wallet-cli/src/device/register-dmk-transport.ts [hunk 1]
```diff
@@ -151,10 +172,12 @@ async function connectFirstUsbDevice(dmk: DeviceManagementKit): Promise<string>
       filter((list: DiscoveredDevice[]) => list.length > 0),
       timeout(CONNECT_TIMEOUT_MS),
     ),
-  );
+  ).catch(() => {
+    throw new Error(NO_LEDGER_DEVICE_FOUND_MESSAGE);
+  });
```
[the remark this answers is not in the document]
Comment (Justkant) [reply apps/wallet-cli/src/device/register-dmk-transport.ts:177]: Acknowledged, but keeping this as-is for now. The broad `.catch()` is intentional here: the "no device" path arrives as a `TimeoutError` (60s) with a real device, but as an `EmptyError` when the discovery stream completes empty (e.g. the `of([])` case covered by the no-device test). So a correct narrowing would have to translate **both** `TimeoutError` and `EmptyError` — only translating timeouts (as suggested) would leak a raw `EmptyError` and break the friendly-message path. Lifecycle-wise nothing leaks: the caller resets the persistent DMK and rethrows on any error regardless. The diagnosability tradeoff is noted and we may revisit, but it's out of scope for this coverage PR.

File: apps/wallet-cli/src/device/register-dmk-transport.test.ts [hunk 2]
```diff
@@ -97,20 +129,196 @@ describe("ensureWalletCliDmkTransport", () => {
...
     await expect(ensureWalletCliDmkTransport()).resolves.toMatchObject({
       sessionId: "session-2",
     });
   });
+
+  it("shares one in-flight connection across concurrent ensure calls", async () => {
+    const connectStarted = deferred();
+    const finishConnect = deferred();
+    const connect = mock(async () => {
+      connectStarted.resolve();
```
[the remark this answers is not in the document]
Comment (Justkant) [reply apps/wallet-cli/src/device/register-dmk-transport.test.ts:228]: False positive — this typechecks fine. `Deferred<void>` makes `resolve` a `(value: void | PromiseLike<void>) => void`, and TypeScript has a special-case rule allowing a `void`-typed parameter to be omitted at the call site, so `resolve()` with no argument is valid (verified with `tsc --strict`). No change needed.

File: apps/wallet-cli/src/device/register-dmk-transport.test.ts [hunk 3]
```diff
@@ -97,20 +129,196 @@ describe("ensureWalletCliDmkTransport", () => {
...
+      return "session-1";
+    });
+    const fake = makeDmk({ connect });
+    const createDeviceManagementKit = mock(async () => fake.kit);
+    createDeviceManagementKitImpl = createDeviceManagementKit;
+
+    const firstPromise = ensureWalletCliDmkTransport();
+    const secondPromise = ensureWalletCliDmkTransport();
+    await connectStarted.promise;
+    finishConnect.resolve();
```
[the remark this answers is not in the document]
Comment (Justkant) [reply apps/wallet-cli/src/device/register-dmk-transport.test.ts:239]: False positive — this typechecks fine. `Deferred<void>` makes `resolve` a `(value: void | PromiseLike<void>) => void`, and TypeScript has a special-case rule allowing a `void`-typed parameter to be omitted at the call site, so `resolve()` with no argument is valid (verified with `tsc --strict`). No change needed.

File: apps/wallet-cli/src/device/register-dmk-transport.test.ts [hunk 4]
```diff
@@ -97,20 +129,196 @@ describe("ensureWalletCliDmkTransport", () => {
...
+      realSetTimeout(handler, 0, ...args)) as typeof globalThis.setTimeout;
+
+    let resetSettled = false;
+    const resetPromise = resetWalletCliDmkSession().then(() => {
+      resetSettled = true;
+    });
+    try {
+      await waitOneTick();
+      expect(resetSettled).toBe(false);
+      finishDestroy.resolve();
```
[the remark this answers is not in the document]
Comment (Justkant) [reply apps/wallet-cli/src/device/register-dmk-transport.test.ts:269]: False positive — this typechecks fine. `Deferred<void>` makes `resolve` a `(value: void | PromiseLike<void>) => void`, and TypeScript has a special-case rule allowing a `void`-typed parameter to be omitted at the call site, so `resolve()` with no argument is valid (verified with `tsc --strict`). No change needed.

File: apps/wallet-cli/src/device/register-dmk-transport.test.ts [hunk 5]
```diff
@@ -97,20 +129,196 @@ describe("ensureWalletCliDmkTransport", () => {
...
+    } as never);
+    await ensureWalletCliDmkTransport();
+
+    let disposeSettled = false;
+    const disposePromise = disposeWalletCliDmkTransportFully().then(() => {
+      disposeSettled = true;
+    });
+    await waitOneTick();
+    expect(disposeSettled).toBe(false);
+    finishDisconnect.resolve();
```
[the remark this answers is not in the document]
Comment (Justkant) [reply apps/wallet-cli/src/device/register-dmk-transport.test.ts:318]: False positive — this typechecks fine. `Deferred<void>` makes `resolve` a `(value: void | PromiseLike<void>) => void`, and TypeScript has a special-case rule allowing a `void`-typed parameter to be omitted at the call site, so `resolve()` with no argument is valid (verified with `tsc --strict`). No change needed.
